### PR TITLE
Feature/add search release

### PR DESCRIPTION
**Pull request by PatrickConnollyONS**
### What

Phase 2 of adding Elastic Search (ES) capabilities for Release (Calendar) entries. These commits:
- Add templates for Elastic Search Release Calendar entry queries
- Add the ReleaseBuilder and logic for parsing and validating the query parameters
- Add a utility to build and/or execute queries against an Elastic search server

### How to review

Check code and ensure all tests are passing

### Who can review

Anyone but someone familiar with dp-search-api would be best placed

### CODE REVIEW

File: query/releasesearch.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	"fmt"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+)
+
+type ParamValidator map[paramName]validator
+
+func (qpv ParamValidator) Validate(name, value string) (interface{}, error) {
```
Comment (rahulPalliyalil) [inline query/releasesearch.go:17]: (question): can this return a specific type rather than a generic interface ?

File: query/releasesearch.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+	if d.After(time.Date(2200, 1, 1, 0, 0, 0, 0, time.UTC)) {
+		return Date{}, errors.New("date too far in future")
+	}
+
+	return Date(d), nil
+}
+
+func MustParseDate(date string) Date {
+	d, e := ParseDate(date)
```
Comment (rahulPalliyalil) [inline query/releasesearch.go:96]: (minor): if you can rename this to ```err``` instead of ```e``` please

File: query/releasesearch_test.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	"testing"
+	"text/template"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+var validators = NewReleaseQueryParamValidator()
+
+func TestLimit(t *testing.T) {
+	Convey("given a limit validator, and a set of limits as strings", t, func() {
```
Comment (rahulPalliyalil) [inline query/releasesearch_test.go:16]: (minor): again i guess these can be run in parallel ? ```t.Parallel()``` ?

File: query/releasesearch_test.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+			v, e := validator(ls.given)
+
+			So(v, ShouldEqual, ls.exValue)
+			So(e, ShouldResemble, ls.exError)
+		}
+	})
+}
+
+func TestOffset(t *testing.T) {
+	Convey("given an offset validator, and a set of offsets as strings", t, func() {
```
Comment (rahulPalliyalil) [inline query/releasesearch_test.go:40]: same as above , can be run in parallel

File: templates/search/releasecalendar/main.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+func main() {
+	var (
+		sr = query.ReleaseSearchRequest{
+			Size:           10,
+			SortBy:         query.RelDateDesc,
+			Term:           "Education in Wales",
+			ReleasedAfter:  query.MustParseDate("2018-01-01"),
+			ReleasedBefore: query.MustParseDate("2018-12-31"),
+			//Upcoming: true,
```
Comment (rahulPalliyalil) [inline templates/search/releasecalendar/main.go:30]: (minor): Can we remove this commented code if this is not used ?

File: templates/search/releasecalendar/main.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		err = json.Compact(&b, uq)
+		if err != nil {
+			log.Error(ctx, "failed to compact query", err)
+			os.Exit(4)
+		}
+		q = b.Bytes()
+	}
+
+	if *multi {
+		q, err = query.FormatMultiQuery(append([]byte(`{"index" : "ons", "type": ["release"], "search_type": "dfs_query_then_fetch"}$$`), append(q, []byte(`$$`)...)...))
```
Comment (rahulPalliyalil) [inline templates/search/releasecalendar/main.go:85]: (suggestion): I would prefer if this ```{"index" : "ons", "type": ["release"], "search_type": "dfs_query_then_fetch"}$$```is extracted to a variable for readability.

File: templates/search/releasecalendar/main.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		if err != nil {
+			log.Error(ctx, "failed to format multi query", err)
+			os.Exit(5)
```
Comment (rahulPalliyalil) [inline templates/search/releasecalendar/main.go:88]: (question): why should this exit with 5 status ? and why can't this be 1 ?

**Review by rahulPalliyalil — CHANGES_REQUESTED** [review]

File: query/releasesearch.go (hunk 1 above)
Comment (PatrickConnollyONS) [reply query/releasesearch.go:17]: not really I think -  the idea is that the validation may produce a value of any type, such as date and `sort in thos case, and the caller knows the type of the parameter and can assert that on the return value.
An alternative of providing an actual specific interface would be not be viable; there is one alternative which is to have a signature of
Validate(name, value string, actualValue interface{}) error
where the caller then provides a pointer to the parameter, like passing a pointer when calling
json.Unmarshal(value []byte, actualValue interface{}) error

Note sure it adds a lot in this instance - what do you think?

File: templates/search/releasecalendar/main.go (hunk 7 above)
Comment (PatrickConnollyONS) [reply templates/search/releasecalendar/main.go:88]: I was using a specific return value to indicate the reason for failure. It's not strictly necessary and I have changed it. The same with the other comments below - this binary was really a quick and dirty tool for me to run queries, but I should have tidied it up a bit more before checking it in

File: query/releasesearch.go (hunk 1 above)
Comment (rahulPalliyalil) [reply query/releasesearch.go:17]: sure, that's fine. We can leave it as it is I guess.

**Review by rahulPalliyalil — APPROVED** [review]

**Pull request merged**